### DISCUSSION TITLE
Consulta Tamagotchi

### DISCUSSION THREAD

**Discussion by Gruschka** [dds-jv/foro#173] [Consultas de objetos] [answered]
Buenas!

Estaba haciendo el [ejercicio Tamagotchi](https://sites.google.com/site/utndesign/material/guia-de-ejercicios/guia-objetos-patrones/tamagotchi), que pide modelar el estado de una mascota virtual y que, según el mismo, las acciones de comer() y jugar() tengan distintos efectos.

Mi duda pasa por la siguiente parte del enunciado:

> Cuando una mascota juega, pasa lo siguiente:
>- Si está contenta, su nivel se incrementa en dos unidades. Si llegará a jugar mas de 5 veces se pone hambrienta (jugar da hambre).
>- Si está de mal humor, se pone contenta.
>-  **_Se pone de mal humor si esta hambrienta._**
>-  **_Una mascota puede jugar si está contenta o aburrida, si está hambrienta no._**

Entiendo que la parte de _aburrida_ es un typo porque sólo ahi aparece. Ahora bien, por un lado dice que se pone de malhumor si come estando hambrienta y luego dice que no puede jugar si está hambrienta, lo cual resulta contradictorio.

Más alla de esta contradicción, yo asumí que no puede jugar estando hambrienta y duda principal gira en torno a como modelar el 'no poder' realizar una accion:

Yo modele el Estado de una mascota como una interfaz, y luego hice una clase por cada estado con la implementacion correspondiente de los metodos comer y jugar. La clase Mascota delegaria a su estado aplicar en si misma los efectos de estas acciones.

Ahora bien, en el caso particular del estado hambriento, qué tendría que hacer para modelar que _no pueda_ jugar? Hacer que el método no haga nada suena raro porque daría la idea de que si puede y que no tiene efecto. Arrojar una exception tampoco me termina de cerrar porque según entendí, la idea es más bien tirar exceptions en código que no depende de uno (tipo integraciones con 3rd parties).

Gracias!

**Comment by gastonprieto** [comment]
Buenas @Gruschka !

Ese enunciado esta viejo, donde lo viste así lo actualizo. El enunciado es el de este apunte: https://docs.google.com/document/d/1RnmKXlQ5tdIasBmFiYFOOe8_JZMSetaT1APLLE9NGIE/edit

---

Igual te respondo tus preguntas:

> Ahora bien, por un lado dice que se pone de malhumor si come estando hambrienta y luego dice que no puede jugar si está hambrienta, lo cual resulta contradictorio.

No entiendo lo contradictorio de esto, uno habra sobre lo que le pasa al comer y el otro habla si puede jugar o no .

--

> yo asumí que no puede jugar estando hambrienta

Precisamente eso es lo que dice el enunciado:

> Una mascota puede jugar si está contenta o aburrida, *si está hambrienta no* 

----

> Ahora bien, en el caso particular del estado hambriento, qué tendría que hacer para modelar que no pueda jugar? Hacer que el método no haga nada suena raro porque daría la idea de que si puede y que no tiene efecto. 

Exacto, podría ser una opción pero deberías ver que significa a nivel negocio, si es que no te da bola o que rompe. Asumamos que debería romper

> Arrojar una exception tampoco me termina de cerrar porque según entendí, la idea es más bien tirar exceptions en código que no depende de uno (tipo integraciones con 3rd parties).

No necesariamente. Las excepciones se usan para comunicar que hay un error, en este caso el error es que alguien mando el mensaje `jugar` cuando esta en un estado que no debía y haces que falle, lanzando una excepción. 
Son mecanismos que se usan en el dominio muchas veces, no solo con cosas de terceros. 

Creo que te estas confundiendo con el concepto de confiar en el adentro y desconfiar en el afuera. Por ejemplo, podes confiar por ejemplo que entre tus objetos no se manda NULL en los parametros por decir algo, o que estan bien construidos porque expusiste un constructor con todas las dependencias que necesita, pero deberías tener menos confianza en el formato de las cosas que te mandan de afuera, por ejemplo ingresos del usuario o respuestas de sistemas externos, en esos hay que tener menos confianza, porque no estan bajo tu control.

Por otro lado creo que te esta faltando un requerimiento:
> Preguntarle si puede jugar

Seguramente alguien antes de mandarte el `mascota.jugar()` como sabe que puede fallar, le mande un mensaje si puede jugar. Es como el validar que un archivo exista antes de abrirlo.

**Comment by Gruschka** [comment, marked as answer]
Hola Gastón, muchas gracias por la pronta respuesta!!

El enunciado lo saque de la [seccion 3 de ejercicios de la pagina](https://dds-jv.github.io/ejercicios/guia-3/) de la cursada, la cual estaba linkeada en la [ultima presentacion ](https://docs.google.com/presentation/d/1nu2ykrWLkdWewDWQ7sP_cdcU9bvKFhaMm2vctFXwe8w/edit#slide=id.g8c51854c30_0_31) de la clase de los viernes.

Me equivoqué al explicar qué me parecía contradictorio. Me refería a que el enunciado decia:

> "...Cuando una mascota juega, pasa lo siguiente: ... Se pone de mal humor si esta hambrienta."

y despues dice:

> Una mascota puede jugar si está contenta o aburrida, si está hambrienta no.

Entonces por un lado dice que si está hambrienta puede jugar, y se pone de mal humor, y por otro lado dice que no puede. De todos modos por lo que veo en el link que me compartis, esto es algo del enunciado viejo.

Me quedó clarísimo lo de las excepciones, es algo que me había quedado medio colgado. Mil gracias!!
